Partially stablize runtime metrics

ranger-ross opened the pull request:
This PR partially stabilizes the runtime metrics feature.

As discussed in #86 , the approach is to stabilize `tokio_metrics::RuntimeMonitor` and the metrics that are considered stable in tokio while keeping unstable metrics opt-in behind the `tokio_unstable` config.

Also while reviewing, you might want to be aware that the [tokio docs](https://docs.rs/tokio/latest/tokio/runtime/struct.RuntimeMetrics.html) on docs.rs are incorrectly showing some metrics as stable when they are not. (See https://github.com/tokio-rs/tokio/issues/6745)

closes #86

ranger-ross commented:
> I can probably knock out a fix for that, unless you were interested in taking it?

I could probably take a look at that later this week or next. :)

ranger-ross commented on src/runtime/metrics_rs_integration.rs:
```diff
@@ -243,61 +246,76 @@ pub struct RuntimeMetricsReporter {
         $(
             #[doc = $doc:tt]
-            $name:ident: $kind:tt <$unit:ident> $opts:tt
+            $(@$unstable:ident)? $name:ident: $kind:tt <$unit:ident> $opts:tt
```
Sure, I can try to do that. Though disclaimer, my macro-fu is a bit weak. :P

ranger-ross commented:
I pushed some changes to address your feedback. lmk what you think.

Also side note: When I was testing my CI changes on my fork i noticed that the `check` job has started failing with the release of Rust 1.88 due to the [uninlined_format_args](https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args) clippy lint being moved from `pedantic` to `style`. A few examples were failing the lint so I went ahead and fixed those.

jlizen commented on src/runtime.rs:
```diff
@@ -51,1120 +51,1184 @@ pub struct RuntimeMonitor {
-    total_polls_count: u64,
-    total_busy_duration: Duration,
-    poll_time_histogram: Vec<u64>,
+define_worker! {
```
Is there a reason that this really needs to define the whole struct rather than just inner fields? That way we could use it for `RuntimeIntervals` fields as well.

For that matter, we could probably get by with just one macro total... just accept another annotation specifying whether the unstable flag needs the docsrs tag or not. Up to you if you care to mess with that. But anyway we will want to apply the macro to at least the `RuntimeIntervals`

jlizen commented on src/runtime.rs:
```diff
@@ -1177,19 +1241,36 @@ pub struct RuntimeIntervals {
     fn probe(&mut self) -> RuntimeMetrics {
         let now = Instant::now();
 
+        #[cfg(tokio_unstable)]
```
I imagine you haven't gotten to this one yet, but still needs similar cleanup. Similar suggestion around instantiating the stable fieldset first with no cfg flag, with `..Default::default()` (if you need a clippy lint override around useless default, that's fine), and then mutate in the unstable ones?

jlizen commented on src/runtime.rs:
```diff
@@ -1352,7 +1451,9 @@ impl Worker {
             }};
         }
 
+        #[cfg(tokio_unstable)]
```
Any reason these two variables couldn't be moved into the unstable block below?

ranger-ross replied:
nope, i merged them together :)

jlizen commented on src/runtime.rs:
```diff
@@ -1304,22 +1391,34 @@ impl RuntimeMonitor {
         };
 
-        Worker {
+        #[cfg(not(tokio_unstable))]
```
This technically works but it's going to snowball over time. Can we `#[derive(Default)]` on worker, and then instantiate it mutably with no config flag and `..Default::default()` (we can override the clippy lint against useless default). And mutate to set the needed fields for unstable? 

It should optimize down to the same thing, but we will only set fields once.

jlizen commented on src/runtime/metrics_rs_integration.rs:
```diff
@@ -243,84 +246,123 @@ pub struct RuntimeMetricsReporter {
-            $name:ident: $kind:tt <$unit:ident> $opts:tt
-        ),*
-        $(,)?
+         stable {
```
This is great, thanks!

jlizen commented on src/runtime.rs:
````diff
@@ -195,6 +195,7 @@ pub struct RuntimeMetrics {
     ///     println!("mean task poll duration is {:?}", interval.mean_poll_duration);
     /// }
     /// ```
+    #[cfg(tokio_unstable)]
````
I'm not actually sure whether this properly inherits like you're suggesting or not... Ideally you're right! If not, it might be worth cutting them an issue to make that the behavior.

ranger-ross commented on src/runtime.rs:
```diff
@@ -51,1120 +51,1184 @@ pub struct RuntimeMonitor {
+            $(
+                $(#[$($unstable_attributes)*])*
+                #[cfg(tokio_unstable)]
+                #[cfg_attr(docsrs, doc(cfg(tokio_unstable)))]
```
I tried setting up docsrs locally and got it running but was getting some strange behavior and couldn't get the `--local` builds working :(

I can probably try again later this week.

ranger-ross replied:
thanks for checking! 
Added the rt feature flag in the [latest commit](https://github.com/tokio-rs/tokio-metrics/compare/73cf76ecc9f1da3983dc67233a9f137d9e2e104a..f1d3e99ad4f86ee29731efb23f9bb84312db5801)

ranger-ross commented:
I pushed some fixes to address your comments. Hopefully I addressed all of your feedback (outside of [this comment](https://github.com/tokio-rs/tokio-metrics/pull/87#discussion_r2175301482))

Let me know if I missed any or you'd like to see other changes

ranger-ross commented:
Regarding the hanging CI, if I recall correctly its because `test-versions` is a "required check" in this repository's settings.

Since I split it into 2 job `test-versions (stable)` and `test-versions (unstable)` the name no longer matches and GitHub is waiting for the `test-versions` job to complete. (which no longer exists)

I tested in this on my fork and was able to reproduce by adding a branch protection rule.
<details><summary>Settings</summary>

![image](https://github.com/user-attachments/assets/9992ee91-cbc0-46ae-883d-85048b7c1dd3)

</details> 


I think you can update the repo settings before merging to add the new CI jobs as required.

jlizen reviewed (approved):
> Regarding the hanging CI, if I recall correctly its because test-versions is a "required check" in this repository's settings.

Makes sense, thanks for the pointer!

Re: the change to add the `rt` docsrs flag - thanks for that. It looks like I misinterpreted the previous state though - I see the fields expanding to show the proper docsrs config, but the generated docs still only show `tokio-unstable` on the various fields. It looks like the docs.rs display itself implicitly scopes the broadest gates to the struct indicator, and then only supplements on fields. You can see the same behavior on the [tokio docs](https://docs.rs/tokio/latest/tokio/runtime/struct.RuntimeMetrics.html).

Anyway we can leave the macro as-is, it's redundant but harmless. Sorry for the confusion.
